### PR TITLE
[native-stack] Add transitionStart and transitionEnd events

### PR DESCRIPTION
**Pull request by janicduplessis**
Implement [these](https://github.com/react-navigation/react-navigation/blob/bea14aa26fd5cbfebc7973733c5cf1f44fd323aa/packages/stack/src/types.tsx#L25) events in native stack, similar to the events available in js stack.

### CODE REVIEW

**Comment by Hirbod** [comment]
Nice one! I miss this events badly!

**Comment by Hirbod** [comment]
@janicduplessis https://github.com/software-mansion/react-native-screens/issues/317
could this PR also take this into consideration? Without these I think react-native-shared-element won't be able to support native-stack

**Comment by janicduplessis** [comment]
@Hirbod I think it is different, It could be possible to expose an animation progress event for the custom animations, so the fade one right now. Not sure if it is possible with the default push / pop animations.

**Comment by Hirbod** [comment]
I think the fade one would be an extremely good start :)

File: src/index.d.ts [hunk 1]
```diff
@@ -29,12 +29,25 @@ declare module 'react-native-screens' {
     active?: 0 | 1 | Animated.AnimatedInterpolation;
     onComponentRef?: (view: any) => void;
     children?: React.ReactNode;
+
+    /**
+     * @description A callback that gets called when the current screen will appears. This is called as soon as the transition begins.
```
Comment (WoLewicki) [inline src/index.d.ts:34]:
```suggestion
     * @description A callback that gets called when the current screen will appear. This is called as soon as the transition begins.
```

File: src/index.d.ts [hunk 2]
```diff
@@ -29,12 +29,25 @@ declare module 'react-native-screens' {
...
+     */
+    onWillAppear?: (e: NativeSyntheticEvent<NativeTouchEvent>) => void;
     /**
-     *@description A callback that gets called when the current screen appears.
+     * @description A callback that gets called when the current screen will disappears. This is called as soon as the transition begins.
```
Comment (WoLewicki) [inline src/index.d.ts:38]:
```suggestion
     * @description A callback that gets called when the current screen will disappear. This is called as soon as the transition begins.
```

**Comment by WoLewicki** [comment]
Could you also update https://github.com/software-mansion/react-native-screens/blob/master/native-stack/README.md with the new events? And I think it will be ready then.

**Comment by janicduplessis** [comment]
Updated in 6a72acd1ea1d25d13aaf9ac607b2ee57ec5a7f35, Also removed finishedTransitioning doc since it was removed in a previous PR.

**Comment by janicduplessis** [comment]
Edit: Needed a bit more work to get nested stacks working on Android. This adds some infra for nested stack to register themselves on their parent screen to allow dispatching events to child stacks.

We already looked up the parent screen to get the fragment manager but didn't keep references around. This keeps the reference to the parent Screen in ScreenStack and registers the child stacks in their parent Screen to allow communicating in both directions between nested stacks.

This is mostly useful to dispatch an event from a screen to its child stacks since traversing down the view hierarchy is not practical.

**Comment by janicduplessis** [comment]
Here are logs on iOS and Android testing side by side with https://gist.github.com/janicduplessis/fd2d7b37b19126824a7c3dbbd9891444

image

https://gist.github.com/janicduplessis/33eb975e15ccfc99d6e29882818eec89

**Comment by WoLewicki** [comment]
The PR looks like it is ready to be merged! Very good job @janicduplessis :tada:

**Pull request merged**